sbi/user_input/user_input_checks.py

```diff
@@ -436,37 +449,61 @@ def process_x_o(
         x_o_dim: number of elements in a single data point.
     """
 
+    # Get unbatched simulated data by sampling from prior and simulator.
+    # cast to tensor for comparison
+    simulated_data_batched = torch.as_tensor(
+        simulator(prior.sample((1,))), dtype=float32
+    )
+
+    if x_shape is not None:
+        x_shape = maybe_add_batch_dim_to_size(x_shape)
+        check_for_possibly_batched_x_shape(x_shape)
+        assert x_shape == simulated_data_batched.shape, (
+            f"Observed data shape ({x_shape}) must match "
+            f"simulator output shape ({simulated_data_batched.shape})."
+        )
+    else:
+        # infer the shape of x from simulation.
+        x_shape = simulated_data_batched.shape
+
+    x_o_dim = x_shape[1:].numel()
+
+    return x_shape, x_o_dim
+
+
+def process_x_o(x_o: Tensor, x_shape: torch.Size) -> Tensor:
+    """Return observed data to sbi's shape and type requirements.
+
+    Args:
+        x_o: Observed data as provided by the user.
+        x_shape: The shape that had either been provided by the user at init or had
+            been inferred from simulations.
+
+    Returns:
+        x_o: observed data with shape corrected for usage in SBI.
+    """
+
     # maybe add batch dimension, cast to tensor
     x_o = atleast_2d(x_o)
     x_o = torch.as_tensor(x_o, dtype=float32)
 
     check_for_possibly_batched_observations(x_o)
+    x_o_shape = x_o.shape
 
-    # Get unbatched simulated data by sampling from prior and simulator.
-    # cast to tensor for comparison
-    simulated_data = torch.as_tensor(
-        simulator(prior.sample((1,))), dtype=float32
-    ).squeeze(0)
-
-    # Get data shape by ommitting the batch dimension.
-    x_o_shape = x_o.shape[1:]
-
-    assert x_o_shape == simulated_data.shape, (
```

Comment:
just to double check: this test is replaced by checking x_o vs. x_shape? and x_shape comes from the user or is inferred from the simulator?

Reply:
Exactly!